opentelemetry-otlp/tests/integration_test/tests/metrics.rs

```diff
@@ -1,23 +1,235 @@
+                        .and_then(|s| s.get("name"))
+                        .and_then(|name| name.as_str())
+                        .map_or(false, |n| n == scope_name)
+                });
+
+                // Keep the resource only if it has any matching `ScopeMetrics`
+                !scope_metrics.is_empty()
+            } else {
+                false
+            }
+        });
+    }
+
+    Ok(filtered_json)
+}
+
+/// Performs setup for metrics tests, including environment setup and data seeding.
+/// This only needs to be done once for the whole test suite
+async fn setup_metrics_test() {
+    let mut done = SETUP_DONE.lock().unwrap();
+    if !*done {
+        println!("Running setup before any tests...");
+        *done = true; // Mark setup as done
+
+        // Make sure the collector container is running
+        start_collector_container().await;
+
+        // Initialise the metrics subsystem
+        _ = init_metrics().await;
+    }
+
+    // Truncate results
+    _ = File::create(RESULT_PATH).expect("it's good");
+}
+
+///
+/// Check that the metrics for the given scope match what we expect. This
+/// includes zeroing out timestamps, which we reasonably expect not to match.
+///
+pub fn validate_metrics_against_results(scope_name: &str) -> Result<()> {
+    // Define the results file path
+    let results_file_path = format!("./expected/metrics/{}.json", scope_name);
+
+    // Fetch the actual metrics for the given scope
+    let actual_metrics = fetch_latest_metrics_for_scope(scope_name)
+        .context(format!("Failed to fetch metrics for scope: {}", scope_name))?;
+
+    // Read the expected metrics from the results file
+    let expected_metrics = {
+        let file = File::open(&results_file_path).context(format!(
+            "Failed to open results file: {}",
+            results_file_path
+        ))?;
+        read_metrics_from_json(file)
+    }?;
+
+    // Compare the actual metrics with the expected metrics
+    MetricsAsserter::new(actual_metrics, expected_metrics).assert();
+
+    Ok(())
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    ///
+    /// JSON doesn't roundtrip through the MetricsData models properly.
+    /// TODO - this is some other issue to look at that this test demonstrates.
+    ///
+    //#[tokio::test]
+    //#[ignore]
+    async fn test_roundtrip_example_data() -> Result<()> {
+        let metrics_in = include_str!("../expected/metrics/test_u64_counter_meter.json");
+        let metrics: MetricsData = serde_json::from_str(metrics_in)?;
+        let metrics_out = serde_json::to_string(&metrics)?;
+
+        println!("{:}", metrics_out);
+
+        let metrics_in_json: Value = serde_json::from_str(metrics_in)?;
+        let metrics_out_json: Value = serde_json::from_str(&metrics_out)?;
+
+        assert_eq!(metrics_in_json, metrics_out_json);
+
+        Ok(())
+    }
+
+    #[tokio::test(flavor = "multi_thread", worker_threads = 4)]
+    async fn test_u64_counter() -> Result<()> {
```

Comment:
@pitoniak32 I tried https://github.com/open-telemetry/opentelemetry-rust/pull/2431/files on top of this, and the reqwest::blocking test passed!
(As mentioned in the other PR, just have to run the metric one alone as logs,traces still use tokio::runtime, so won't work)